src/test/java/com/google/research/bleth/utils/QueriesIT.java

```diff
@@ -352,8 +370,165 @@ public void provideNonExistingForeignKey_shouldRetrieveEmptyList() {
+        firstEntityPropertiesValues.put(PROPERTY_A, 1.0);
+        firstEntityPropertiesValues.put(PROPERTY_B, 1.5);
+        firstEntityPropertiesValues.put(PROPERTY_C, 2.0);
+        writeEntityWithProperties(ENTITY_KIND, firstEntityPropertiesValues);
+
+        Map<String, Double> secondEntityPropertiesValues = new HashMap<>();
+        secondEntityPropertiesValues.put(PROPERTY_A, 0.0);
+        secondEntityPropertiesValues.put(PROPERTY_B, 1.5);
+        secondEntityPropertiesValues.put(PROPERTY_C, 0.0);
+        writeEntityWithProperties(ENTITY_KIND, secondEntityPropertiesValues);
+
+        Map<String, Double> thirdEntityPropertiesValues = new HashMap<>();
+        thirdEntityPropertiesValues.put(PROPERTY_A, 2.0);
+        thirdEntityPropertiesValues.put(PROPERTY_B, 1.5);
+        thirdEntityPropertiesValues.put(PROPERTY_C, 4.0);
+        writeEntityWithProperties(ENTITY_KIND, thirdEntityPropertiesValues);
+
+        List<Entity> entities = retrieveEntities(ENTITY_KIND);
+        Set<String> properties = new HashSet<>(Arrays.asList(PROPERTY_A, PROPERTY_B, PROPERTY_C));
+        Map<String, Double> expectedResult = new HashMap<>();
+        expectedResult.put(PROPERTY_A, 1.0);
+        expectedResult.put(PROPERTY_B, 1.5);
+        expectedResult.put(PROPERTY_C, 2.0);
+
+        Map<String, Double> actualResult = Queries.Average(entities, properties);
+
+        assertThat(actualResult).containsExactlyEntriesIn(expectedResult);
+    }
+
+    @Test
+    public void writeMultipleEntitiesWithMissingProperties_shouldCalculateAverageAsExpected() {
+        // Write multiple entities with multiple properties/
+        Map<String, Double> firstEntityPropertiesValues = new HashMap<>();
+        firstEntityPropertiesValues.put(PROPERTY_A, 1.0);
+        writeEntityWithProperties(ENTITY_KIND, firstEntityPropertiesValues);
+
+        Map<String, Double> secondEntityPropertiesValues = new HashMap<>();
+        secondEntityPropertiesValues.put(PROPERTY_A, 0.0);
+        secondEntityPropertiesValues.put(PROPERTY_B, 1.5);
+        writeEntityWithProperties(ENTITY_KIND, secondEntityPropertiesValues);
+
+        Map<String, Double> thirdEntityPropertiesValues = new HashMap<>();
+        thirdEntityPropertiesValues.put(PROPERTY_A, 2.0);
+        thirdEntityPropertiesValues.put(PROPERTY_B, 1.5);
+        writeEntityWithProperties(ENTITY_KIND, thirdEntityPropertiesValues);
+
+        List<Entity> entities = retrieveEntities(ENTITY_KIND);
+        Set<String> properties = new HashSet<>(Arrays.asList(PROPERTY_A, PROPERTY_B, PROPERTY_C));
+        Map<String, Double> expectedResult = new HashMap<>();
+        expectedResult.put(PROPERTY_A, 1.0);
+        expectedResult.put(PROPERTY_B, 1.5);
+        expectedResult.put(PROPERTY_C, Double.NaN);
+
+        Map<String, Double> actualResult = Queries.Average(entities, properties);
+
+        assertThat(actualResult).containsExactlyEntriesIn(expectedResult);
+    }
+
     @After
```

Comment:
Please be consistent: either leave this at the end of the class, or move it to the beginning.
Don't leave it in the middle.